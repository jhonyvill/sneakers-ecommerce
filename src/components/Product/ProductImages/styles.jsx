import styled from "styled-components";

export const Container = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1.5rem;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    height: 50vh;
  }
  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    height: 40vh;
  }
`;

export const FeaturedImageContainer = styled.div`
  position: relative;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    width: 100vw;
    height: 100%;

    &.small{
      width: 90vw;
      height: initial;
      & img {
        border-radius: 10px;
      }
    }
  }
`;

export const FeaturedImage = styled.img`
  border-radius: 10px;
  cursor: pointer;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    border-radius: 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: center;
  }
`;

export const ImagesGrid = styled.div`
  max-width: 27rem;
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  justify-self: center;
  gap: 1.5rem;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    display: none;
  }
`;

export const Image = styled.div`
  position: relative;
  cursor: pointer;

  &::before {
    content: "";
    position: absolute;
    opacity: 0;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    background-color: rgba(255, 255, 255, 0.7);
    transition: opacity 0.1s;
    display: inline-block;
    border-radius: 10px;
    box-sizing: border-box;
  }

  &:hover::before {
    opacity: 1;
  }
  &.active::before {
    border: 2px solid ${({ theme }) => theme.colors.primaryOrange};
    opacity: 1;
  }
  & img {
    flex: 1;
    border-radius: 10px;
  }
`;

export const ButtonWrapper = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
`;

export const ModalButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;

  height: 3rem;
  width: 3rem;
  border-radius: 50%;
  padding: 0.8rem;
  background-color: ${({ theme }) => theme.colors.lightGrayishBlue};
  border: 1px solid transparent;

  & svg {
    scale: 0.8;
  }

  &:hover svg > * {
    stroke: ${({ theme }) => theme.colors.primaryOrange};
  }

  @media (max-width: ${({ theme }) => theme.screen.smallTablet}) {
    padding: 0.6rem;
    & svg {
      scale: 0.7;
    }
  }
  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    height: 2rem;
    width: 2rem;
  }
`;

export const PreviousButton = styled(ModalButton)`
  margin-left: -1.4rem;
  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    margin-left: 1rem;
  }
`;
export const NextButton = styled(ModalButton)`
  margin-right: -1.4rem;
  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    margin-right: 1rem;
  }
`;
