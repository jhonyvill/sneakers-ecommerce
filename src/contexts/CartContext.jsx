import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
