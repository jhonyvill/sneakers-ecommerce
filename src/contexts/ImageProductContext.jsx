import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
