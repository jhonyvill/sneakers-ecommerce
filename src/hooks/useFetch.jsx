import { useCallback, useState } from "react";

export const useFetch = () => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  const request = useCallback(async (url) => {
    let response;
    let json;
    try {
      setError(null);
      setLoading(true);

      response = await fetch(url);
      json = await response.json();
      if (!response.ok) throw new Error(json.message);
    } catch (error) {
      json = null;
      setError(error.message);
      console.log(error.message);
    } finally {
      setData(json);
      setLoading(false);
    }
  }, []);

  return {
    data,
    error,
    loading,
    request,
  };
};
