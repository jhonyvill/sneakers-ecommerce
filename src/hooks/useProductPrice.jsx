import React from "react";

export const useProductPrice = (product) => {
  const discountPercent = product.discount * 100;
  const price = product.price.toFixed(2);
  const discountValue = price * product.discount;
  const finalPrice = (price - discountValue).toFixed(2);
  const hasDiscount = discountPercent > 0;

  function getTotalItemPrice(quantity) {
    return (finalPrice * quantity).toFixed(2);
  }

  return { price, finalPrice, discountPercent, hasDiscount, getTotalItemPrice };
};
