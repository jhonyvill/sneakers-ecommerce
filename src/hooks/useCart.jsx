import React from "react";
import { useState, useContext } from "react";
import { CartContext } from "../contexts/CartContext";

export const useCart = () => {
  const { cartState, cartDispatch } = useContext(CartContext);
  const [quantitySelected, setQuantitySelected] = useState(0);
  const cartHasItems = cartState.items.length > 0;

  function findProductInCart(product) {
    return cartState.items.find((item) => item.product.id === product.id);
  }

  function addCartProduct(product) {
    const existentProduct = findProductInCart(product);
    if(existentProduct){
      cartDispatch({
        type: "UPDATE_ITEM",
        payload: {
          product: product,
          quantity: (existentProduct.quantity += quantitySelected),
        },
      });
    }else if(quantitySelected > 0){
      cartDispatch({
        type: "ADD_ITEM",
        payload: { product: product, quantity: quantitySelected },
      });
    }
    setQuantitySelected(0);
  }

  return {
    quantitySelected,
    setQuantitySelected,
    addCartProduct,
    cartHasItems
  };
};
