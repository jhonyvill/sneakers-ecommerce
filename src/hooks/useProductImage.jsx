import React from "react";
import { useContext } from "react";
import { useState } from "react";
import { ImageProductContext } from "../contexts/ImageProductContext";

export const useProductImage = (images) => {
  const { imageSelected, indexImageSelected, handleSelectedModalImage } = useContext(ImageProductContext);
  const [featuredImage, setFeaturedImage] = useState({
    image: imageSelected || images[0],
    index: indexImageSelected,
  });

  function handleChangeImage(image, indexClicked) {
    indexClicked >= 0 &&
      indexClicked <= images.length - 1 &&
      indexClicked !== featuredImage.index &&
      setFeaturedImage({ image, index: indexClicked });
  }
  function handleClickFeaturedImage() {
    handleSelectedModalImage(featuredImage.image, featuredImage.index);
  }

  return {
    featuredImage: featuredImage.image,
    imageIndexActive: featuredImage.index,
    handleChangeImage,
    handleClickFeaturedImage,
  };
};
