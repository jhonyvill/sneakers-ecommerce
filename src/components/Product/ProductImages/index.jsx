import React from "react";
import * as S from "./styles";
import NextSvg from "../../../assets/icon-next.svg?react";
import PreviousSvg from "../../../assets/icon-previous.svg?react";
import PropTypes from "prop-types";
import { useMedia } from "../../../hooks/useMedia";
import { useProductImage } from "../../../hooks/useProductImage";

const ProductImages = ({ images, modalActive, handleModalActive }) => {
  const {
    featuredImage,
    imageIndexActive,
    handleChangeImage,
    handleClickFeaturedImage,
  } = useProductImage(images);

  const smallScreen = useMedia("(max-width: 820px)");
  
  return (
    <S.Container>
      <S.FeaturedImageContainer
        $container={(smallScreen && modalActive) ? "small" : "normal"}
      >
        <S.FeaturedImage
          width="1000px"
          height="1000px"
          src={featuredImage.src}
          alt={featuredImage.alt}
          onClick={() => {
            if (!modalActive) {
              handleClickFeaturedImage();
              handleModalActive(true);
            }
          }}
        />

        {(modalActive || smallScreen) && (
          <S.ButtonWrapper>
            <S.PreviousButton
              onClick={() => {
                const previousIndex = imageIndexActive - 1;
                handleChangeImage(images[previousIndex], previousIndex);
              }}
            >
              <PreviousSvg />
            </S.PreviousButton>
            <S.NextButton
              onClick={() => {
                const nextIndex = imageIndexActive + 1;
                handleChangeImage(images[nextIndex], nextIndex);
              }}
            >
              <NextSvg />
            </S.NextButton>
          </S.ButtonWrapper>
        )}
      </S.FeaturedImageContainer>

      <S.ImagesGrid>
        {images.map((image, index) => {
          return (
            <S.Image
              key={index}
              className={imageIndexActive === index && "active"}
              onClick={() => handleChangeImage(image, index)}
            >
              <img
                width="176px"
                height="176px"
                src={image.thumb}
                alt={image.alt}
              />
            </S.Image>
          );
        })}
      </S.ImagesGrid>
    </S.Container>
  );
};

ProductImages.propTypes = {
  images: PropTypes.array.isRequired,
  modalActive: PropTypes.bool,
  handleModalActive: PropTypes.func,
};

export default ProductImages;
