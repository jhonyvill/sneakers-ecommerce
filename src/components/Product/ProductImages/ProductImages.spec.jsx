import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme, productDefault } from "../../../../.jest/utils";
import { ImageProductProvider } from "../../../contexts/ImageProductContext";
import * as hookProductImage from "../../../hooks/useProductImage";
import * as hookMedia from "../../../hooks/useMedia";
import ProductImages from ".";

const images = productDefault.pictures;

function renderProductImages(modalActive, mockFunction = jest.fn()) {
  return renderWithTheme(
    <ImageProductProvider>
      <ProductImages
        images={images}
        modalActive={modalActive}
        handleModalActive={mockFunction}
      />
    </ImageProductProvider>
  );
}

function createUseProductImageSpy() {
  return jest.spyOn(hookProductImage, "useProductImage").mockReturnValue({
    featuredImage: images[0],
    imageIndexActive: 0,
    handleChangeImage: jest.fn(),
    handleClickFeaturedImage: jest.fn(),
  });
}

function createUseMediaSpy(smallScreen) {
  return jest.spyOn(hookMedia, "useMedia").mockReturnValue(smallScreen);
}

jest.mock("../../../hooks/useMedia", () => ({
  __esModule: true,
  useMedia: () => {
    return false;
  },
}));

describe("<ProductImages />", () => {
  it("should render a ProductImages component with featured image and images", () => {
    const { getByTestId } = renderProductImages(false);

    const featuredImage = getByTestId("featured-image");
    const totalImagesOnPage = screen.getAllByRole("img");

    expect(featuredImage).toBeInTheDocument();
    expect(totalImagesOnPage.length).toBe(5);
  });

  it("should render next and previous buttons when 'modalActive' property is true", () => {
    renderProductImages(true);

    const previousButton = screen.getByRole("button", {
      name: /imagem anterior/i,
    });
    const nextButton = screen.getByRole("button", { name: /próxima imagem/i });

    expect(previousButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
  });

  it("should render next and previous buttons when 'smallScreen' property is true", () => {
    const spy = createUseMediaSpy(true);
    renderProductImages(false);

    const previousButton = screen.getByRole("button", {
      name: /imagem anterior/i,
    });
    const nextButton = screen.getByRole("button", { name: /próxima imagem/i });

    expect(previousButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
    spy.mockRestore();
  });

  it("should render featured image with className 'small' when 'modalActive' and 'smallScreen' are true", () => {
    createUseMediaSpy(true);
    const { getByTestId } = renderProductImages(true);
    const featuredImageContainer = getByTestId("container-featured-image");

    expect(featuredImageContainer).toHaveClass("small");
  });

  it("should call two functions on click featured image if 'modalActive' prop is false", async () => {
    const spy = createUseProductImageSpy();

    const handleModalActiveMock = jest.fn();
    const { getByTestId } = renderProductImages(false, handleModalActiveMock);
    const featuredImage = getByTestId("featured-image");

    await userEvent.click(featuredImage);

    expect.assertions(2);
    expect(spy().handleClickFeaturedImage).toHaveBeenCalledTimes(1);
    expect(handleModalActiveMock).toHaveBeenCalledTimes(1);
    spy.mockRestore();
  });

  it("shouldn't call two functions on click featured image if 'modalActive' prop is true", async () => {
    const spy = createUseProductImageSpy();

    const handleModalActiveMock = jest.fn();
    const { getByTestId } = renderProductImages(true, handleModalActiveMock);
    const featuredImage = getByTestId("featured-image");

    await userEvent.click(featuredImage);

    expect.assertions(2);
    expect(spy().handleClickFeaturedImage).toHaveBeenCalledTimes(0);
    expect(handleModalActiveMock).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });
});

describe("<ProductImages /> snapshots", () => {
  it("should match snapshot", () => {
    createUseMediaSpy(false);
    const { container } = renderProductImages(false);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when small screen", () => {
    createUseMediaSpy(true);
    const { container } = renderProductImages(false);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when modal active", () => {
    createUseMediaSpy(false);
    const { container } = renderProductImages(true);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when modal active and small screen", () => {
    createUseMediaSpy(true);
    const { container } = renderProductImages(true);
    expect(container.firstChild).toMatchSnapshot();
  });
});
