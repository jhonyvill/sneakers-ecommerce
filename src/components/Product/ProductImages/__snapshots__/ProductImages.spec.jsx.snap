// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImages /> snapshots should match snapshot 1`] = `
<section
  aria-label="imagens do produto"
  class="sc-bdnyFh fPwIWZ"
>
  <div
    class="sc-gtsqUy ihcXiq"
    data-testid="container-featured-image"
  >
    <img
      alt="product 1"
      class="sc-dlniIP iqruGc"
      data-testid="featured-image"
      height="1000px"
      src="image-1.jpg"
      width="1000px"
    />
  </div>
  <div
    class="sc-hKFymg ljGcIY"
  >
    <div
      class="sc-eCAqax ddLQhb active"
      data-testid="grid-image-0"
    >
      <img
        alt="product 1"
        height="176px"
        src="thumbnail-1.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-1"
    >
      <img
        alt="product 2"
        height="176px"
        src="thumbnail-2.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-2"
    >
      <img
        alt="product 3"
        height="176px"
        src="thumbnail-3.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-3"
    >
      <img
        alt="product 4"
        height="176px"
        src="thumbnail-4.jpg"
        width="176px"
      />
    </div>
  </div>
</section>
`;

exports[`<ProductImages /> snapshots should match snapshot when modal active 1`] = `
<section
  aria-label="imagens do produto"
  class="sc-bdnyFh fPwIWZ"
>
  <div
    class="sc-gtsqUy ihcXiq"
    data-testid="container-featured-image"
  >
    <img
      alt="product 1"
      class="sc-dlniIP iqruGc"
      data-testid="featured-image"
      height="1000px"
      src="image-1.jpg"
      width="1000px"
    />
    <div
      class="sc-jSFipO cYlmsE"
    >
      <button
        aria-label="imagem anterior"
        class="sc-gKAaef sc-iCoHzw jGlZmE jPrXqN"
      >
        <icon-mock />
      </button>
      <button
        aria-label="próxima imagem"
        class="sc-gKAaef sc-fujznN jGlZmE gzxJie"
      >
        <icon-mock />
      </button>
    </div>
  </div>
  <div
    class="sc-hKFymg ljGcIY"
  >
    <div
      class="sc-eCAqax ddLQhb active"
      data-testid="grid-image-0"
    >
      <img
        alt="product 1"
        height="176px"
        src="thumbnail-1.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-1"
    >
      <img
        alt="product 2"
        height="176px"
        src="thumbnail-2.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-2"
    >
      <img
        alt="product 3"
        height="176px"
        src="thumbnail-3.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-3"
    >
      <img
        alt="product 4"
        height="176px"
        src="thumbnail-4.jpg"
        width="176px"
      />
    </div>
  </div>
</section>
`;

exports[`<ProductImages /> snapshots should match snapshot when modal active and small screen 1`] = `
<section
  aria-label="imagens do produto"
  class="sc-bdnyFh fPwIWZ"
>
  <div
    class="sc-gtsqUy ihcXiq small"
    data-testid="container-featured-image"
  >
    <img
      alt="product 1"
      class="sc-dlniIP iqruGc"
      data-testid="featured-image"
      height="1000px"
      src="image-1.jpg"
      width="1000px"
    />
    <div
      class="sc-jSFipO cYlmsE"
    >
      <button
        aria-label="imagem anterior"
        class="sc-gKAaef sc-iCoHzw jGlZmE jPrXqN"
      >
        <icon-mock />
      </button>
      <button
        aria-label="próxima imagem"
        class="sc-gKAaef sc-fujznN jGlZmE gzxJie"
      >
        <icon-mock />
      </button>
    </div>
  </div>
  <div
    class="sc-hKFymg ljGcIY"
  >
    <div
      class="sc-eCAqax ddLQhb active"
      data-testid="grid-image-0"
    >
      <img
        alt="product 1"
        height="176px"
        src="thumbnail-1.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-1"
    >
      <img
        alt="product 2"
        height="176px"
        src="thumbnail-2.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-2"
    >
      <img
        alt="product 3"
        height="176px"
        src="thumbnail-3.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-3"
    >
      <img
        alt="product 4"
        height="176px"
        src="thumbnail-4.jpg"
        width="176px"
      />
    </div>
  </div>
</section>
`;

exports[`<ProductImages /> snapshots should match snapshot when small screen 1`] = `
<section
  aria-label="imagens do produto"
  class="sc-bdnyFh fPwIWZ"
>
  <div
    class="sc-gtsqUy ihcXiq"
    data-testid="container-featured-image"
  >
    <img
      alt="product 1"
      class="sc-dlniIP iqruGc"
      data-testid="featured-image"
      height="1000px"
      src="image-1.jpg"
      width="1000px"
    />
    <div
      class="sc-jSFipO cYlmsE"
    >
      <button
        aria-label="imagem anterior"
        class="sc-gKAaef sc-iCoHzw jGlZmE jPrXqN"
      >
        <icon-mock />
      </button>
      <button
        aria-label="próxima imagem"
        class="sc-gKAaef sc-fujznN jGlZmE gzxJie"
      >
        <icon-mock />
      </button>
    </div>
  </div>
  <div
    class="sc-hKFymg ljGcIY"
  >
    <div
      class="sc-eCAqax ddLQhb active"
      data-testid="grid-image-0"
    >
      <img
        alt="product 1"
        height="176px"
        src="thumbnail-1.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-1"
    >
      <img
        alt="product 2"
        height="176px"
        src="thumbnail-2.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-2"
    >
      <img
        alt="product 3"
        height="176px"
        src="thumbnail-3.jpg"
        width="176px"
      />
    </div>
    <div
      class="sc-eCAqax ddLQhb"
      data-testid="grid-image-3"
    >
      <img
        alt="product 4"
        height="176px"
        src="thumbnail-4.jpg"
        width="176px"
      />
    </div>
  </div>
</section>
`;
